### PR TITLE
8333748: javap crash - Fatal error: Unmatched bit position 0x2 for location CLASS

### PR DESCRIPTION
**Pull request by liach**
Currently, javap crashes for class files that have set non-zero values for undefined access flag bits, as `java.lang.reflect.AccessFlag.maskToAccessFlag` and `java.lang.classfile.AccessFlags.flags` fail. In contrast, the JVMS, though asking for these bits to be set to 0, requires VM to proceed and ignore these bits. javap should emulate the VM behavior and proceed rendering, ignoring these undefined bits.

In addition, a few bytecode generation utilities in the JDK set unused bits, such as in `java.lang.invoke.MethodHandleImpl.BindCaller#generateInvokerTemplate` and `java.lang.invoke.GenerateJLIClassesHelper#generateCodeBytesForLFs`. Those can be updated in a later cleanup.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8334406](https://bugs.openjdk.org/browse/JDK-8334406) to be approved

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8333748: javap crash - Fatal error: Unmatched bit position 0x2 for location CLASS`

### Issues
 * [JDK-8333748](https://bugs.openjdk.org/browse/JDK-8333748): javap crash - Fatal error: Unmatched bit position 0x2 for location CLASS (**Bug** - P2)
 * [JDK-8334406](https://bugs.openjdk.org/browse/JDK-8334406): javap crash - Fatal error: Unmatched bit position 0x2 for location CLASS (**CSR**)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19708/head:pull/19708` \
`$ git checkout pull/19708`

Update a local copy of the PR: \
`$ git checkout pull/19708` \
`$ git pull https://git.openjdk.org/jdk.git pull/19708/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19708`

View PR using the GUI difftool: \
`$ git pr show -t 19708`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19708.diff">https://git.openjdk.org/jdk/pull/19708.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19708#issuecomment-2166447935)

### CODE REVIEW

**Comment by asotona** [comment]
`javap` should never silently ignore erroneous class file content.
If the flag value violates JVMS - it should be reported as an error and `javap` should reflect that in the return value.
On the other hand `javap` should handle such situations, provide relevant error message and continue with the report.
It is similar case as `javap` detection of inappropriate CP entry types.

File: test/langtools/tools/javap/UndefinedAccessFlagTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+        Files.write(Path.of("transformed.class"), bytes);
+
+        var lines = new JavapTask(toolBox)
+            .classes("transformed.class")
+            .options("-c", "-p", "-v")
+            .run(location == TestLocation.INNER_CLASS ? Task.Expect.SUCCESS : Task.Expect.FAIL)
+            .writeAll()
+            .getOutputLines(Task.OutputKind.DIRECT);
+        assertTrue(lines.stream().anyMatch(l -> l.contains("Unmatched bit position")), () -> String.join("\n", lines));
```
Comment (asotona) [inline test/langtools/tools/javap/UndefinedAccessFlagTest.java:105]: I would add a check the "fatal" error does not occur or any other way to confirm correct recovery.

File: src/jdk.jdeps/share/classes/com/sun/tools/javap/BasicWriter.java [hunk 2]
```diff
@@ -46,6 +72,20 @@ protected BasicWriter(Context context) {
...
+    protected Set<AccessFlag> flagsReportUnknown(AccessFlags flags) {
+        return maskToAccessFlagsReportUnknown(flags.flagsMask(), flags.location());
+    }
+
+    protected Set<AccessFlag> maskToAccessFlagsReportUnknown(int mask, AccessFlag.Location location) {
+        try {
+            return AccessFlag.maskToAccessFlags(mask, location);
+        } catch (IllegalArgumentException ex) {
+            mask &= LOCATION_MASKS.get(location);
+            report(ex);
```
Comment (asotona) [inline src/jdk.jdeps/share/classes/com/sun/tools/javap/BasicWriter.java:84]: Unfortunately the original exception message is missing any info that it is related to access flags and it is not very clear what "Error: Unmatched bit position 0x2 for location CLASS" means.
I would add at least a message prefix, for example "Error: Access Flags: Unmatched bit position 0x2 for location CLASS".

Comment (liach) [reply src/jdk.jdeps/share/classes/com/sun/tools/javap/BasicWriter.java:84]: Agreed, even though this message is already printed right by the access modifiers.

File: test/langtools/tools/javap/UndefinedAccessFlagTest.java (hunk 1 above)
Comment (liach) [reply test/langtools/tools/javap/UndefinedAccessFlagTest.java:105]: Great suggestion! I will filter the lines starting with `Error:` and assert they `allMatch(st -> st.contains("Access Flags:"))`

**Comment by jddarcy** [comment]
@liach , please file a CSR for this issue to discuss the behavioral changes.

**Review by asotona — APPROVED** [review]
Now it looks good to me, thanks.

**Pull request closed without merging**